docs/proposals/application-name-identifier.md

```diff
@@ -0,0 +1,397 @@
+label_ to the name of the managing `Application` on all resources that are
+managed (i.e. reconciled from Git). The default label used is the well-known
+label `app.kubernetes.io/instance`.
+
+This proposal suggests to change the _value_ of this label to not use the
+literal application name, but instead use a value from a stable and reasonable
+collision free hash algorithm.
+
+## Motivation
+
+The main motivation behind this change is the Kubernetes restriction for the
+maximum allowed length of label values, which no more than 63 characters.
+
+In large scale installations, in order to build up an easy to understand and
+well-formed naming schemes for applications managed by Argo CD, people often
+hit the 63 character limit and need to define the naming scheme around this
+unnecessary limit.
+
+Furthermore, proposed changes such as described in
+https://github.com/argoproj/argo-cd/pull/6409
+would require the application names to include more implicit information
+(such as the application's source namespace), which will even add more
+characters to the application names.
+
+An additional motivation - while we're at touching at application instance
+label - is to improve the way how multiple Argo CD instances could manage
+applications on the same cluster, without requiring the user to actually
+perform instance specific configuration.
+
+### Goals
+
+* Allow application names of more than 63 characters
+
+* Keep using a label to properly _select_ resources managed by an Application
+  using Kubernetes label selectors
+
+* Keep having a human-readable way to identify resources that belong to a
+  given Argo CD application
+
+* As a stretch-goal, allow multiple Argo CD instances to manage resources on
+  the same cluster without the need for configuring application instance label
+  key (usually `app.kubernetes.io/instance`)
+
+### Non-Goals
+
+* Change the default name of the application instance label
+
+## Proposal
+
+We propose to move from a _human-readable_ value to identify the resources of
+an application to a _machine-readable_ value of fixed length. For this purpose
+we propose to use a one-way hashing algorithm. The chosen algorithm should be
+_collision-free_ to a certain extent, but does not require to be _secure_ in a
+cryptographic evaluation.
+
+We further propose to add an _annotation_ to the managed resources, which will
+contain the plain text value of the application's name so that humans and other
+consumers of the resources will still be able to identify the application that
+is managing the resource. Annotations in Kubernetes don't suffer from the same
+length restrictions as labels, and can easily store about every possible
+application name that users might come up with.
+
+Furthermore, we may want to consider encoding a unique, persistant identifier
+of the Argo CD installation into the final hash value, e.g.
+
+```
+value = hashFunc(installationIdentifier + "." + applicationName)
```

Comment:
My suggestion would be that hashValue can be applicationName+"."+resourceKey. So that we know what the name of original resourceName in case when some kubernetes operator clone the resource. 

By that we can also aim to solve another usecase of ignoreExtraneous resources outside target namespace

Reply:
I really like the idea to bring in `resourceKey` somewhere. But I think if we put it into the _application instance label_, we will end up with a different value on each resource. This makes it utterly hard to use the label as a selector, e.g. to select all resources belonging to a given application by running `kubectl get somekind -l app.kubernetes.io/instance=<somevalue>`

Maybe it would make more sense to add it as an _additional_ annotation on the resource? E.g. `argo-cd.argoproj.io/resourceKey` instead of trying to condense it.

But at the end, this will be a design decision: Do we want to support above use-case, e.g. using the label as a selector for resources managed by a given application, or not. I clearly see the benefits of using the label as a selector, but I don't have any strong opinions on it. We should just be aware of the consequences.

Reply:
I agree that it is very convenient to be compatible with `kubectl get -l app.kubernetes.io/instance=<somevalue>` . This works most of the time except few edge cases:

* app name is too long
* something else add  `app.kubernetes.io/instance` label and confuses argo cd. This is the most annoying issue and we don't have a very good resolution. 

So I think it would be great if keep using `app.kubernetes.io/instance` label by default and allow to switch to new way when necessary.

@jessesuen had a good point about trying to keep app mark human readable. How about we add new app-level field `resourceTracking` that might be `label` or `annotation`. If user select annotation that Argo CD should apply `app.kubernetes.io/instance` annotation where annotation value is `<appName>-<resouceName>` ?

This would allow us to upgrade smoothly and takes care about both edge cases. WDYT?

Reply:
The `argo-cd.argoproj.io/resourceKey` annotation instead of embedding resource key into `app.kubernetes.io/instance` works too